example/extensions/lib_subgraph/subgraph_lib.cc

```diff
@@ -0,0 +1,278 @@
+      // add output tensor to data storage
+      data.push_back(tmp);
+    } else if (op.compare("exp") == 0) {
+      // get input tensor based on node ID inputs from data storage
+      MXTensor &input = data[node_inputs.list[0].list[0].num];
+      // create temporary storage
+      MXTensor tmp(malloc(input.size()*4), input.shape, input.dtype);
+      // save allocated ptr to free later
+      to_free.push_back(tmp.data_ptr);
+      // execute exp operator 
+      myExp(input,tmp);
+      // add output tensor to data storage
+      data.push_back(tmp);
+    } else {
+      std::cout << "Error! Unsupported op '" << op << "' found in myExecutor";
+      // free allocated temporary storage
+      for (void* ptr : to_free)
+        free(ptr);
+      return MX_FAIL;
+    }
+  }
+  
+  // get list of outputs from subgraph
+  JsonVal heads = json_val.map[JsonVal("heads")];
+  // copy all operator results to outputs of subgraph
+  for (int j = 0; j < heads.list.size(); j++) {
+    // get computed result
+    MXTensor &result = data[heads.list[0].list[0].num];
+    // get output tensor to pass to MX
+    MXTensor &out = outputs[j];
+    float *out_data = out.data<float>();
+    float *res_data = result.data<float>();
+    // loop and copy data
+    for (int64_t i = 0; i < result.size(); i++) {
+      out_data[i] = res_data[i];
+    }
+  }
+
+  // free allocated temporary storage
+  for (void* ptr : to_free) {
+    free(ptr);
+  }
+  
+  return MX_SUCCESS;
+}
+
+class MyStatefulOp : public CustomStatefulOp {
+ public:
+  explicit MyStatefulOp(std::string sym) : subgraph_sym(sym) {}
+
+  MXReturnValue Forward(std::vector<MXTensor> inputs,
+                        std::vector<MXTensor> outputs,
+                        OpResource op_res) {
+    return myExecutor(inputs, outputs, subgraph_sym);
+  }
+
+ private:
+  std::string subgraph_sym;
+};
+
+MXReturnValue createOpState(std::map<std::string, std::string> attrs,
+                            CustomStatefulOp** op_inst) {
+  std::string serialized_subgraph = "[empty]";
+  // MXNet subgraph is stored as Symbol in operator node attrs subgraphs field
+  // custom subgraph is stored as json string in custom operator attrs map entry
+  if (attrs.count(SUBGRAPH_SYM_JSON)) {
+    // user can now parse json and run other custom ops inside subgraph
+    serialized_subgraph = attrs[SUBGRAPH_SYM_JSON];
+  }
+  *op_inst = new MyStatefulOp(serialized_subgraph);
+  std::cout << "Info: stateful operator created" << std::endl;
+  return MX_SUCCESS;
+}
+
+REGISTER_OP(_custom_subgraph_op)
+.setParseAttrs(parseAttrs)
+.setInferType(inferType)
+.setInferShape(inferShape)
+.setCreateOpState(createOpState);
+
+const std::vector<std::string> op_names({"exp","log"});
+
+MXReturnValue mySupportedOps(std::string json,
+                             const int num_ids,
+                             int *ids,
+                             std::map<std::string, std::string> options) {
+  for (auto kv : options) {
```

Comment:
Should we keep the for loop here?

Reply:
This is part of the user's external library (not MXNet), they may choose to print it or not.

Reply:
yes, its a good example to show users how to create external subgraph libraries and use options users pass